storage/engine: extract static library to fix windows ccl builds

benesch opened the pull request:
Since the start of our Windows support, we've been unable to build CCL binaries because ccl/storageccl/engineccl depended upon symbols from storage/engine. These symbols were not provided when engineccl was built, but rather when the final cockroach binary was linked. On Unix-like platforms, we worked around the issue by passing `-Wl,undefined` or its equivalent to ignore undefined symbols. Windows, however, had no equivalent flag (the Portable Executable format that Windows uses requires all symbols to be resolved at link time), and so we were forced to omit the broken engineccl package by building the OSS binary.

This commit extracts the C and C++ code from engine and engineccl into their own static libraries, libroach and libroachccl, that are built alongside our third-party C and C++ dependencies. This means engineccl simply links against libroach *and* libroachccl so that all symbols are available at build time. (Actually, there are two Go functions in storage/engine that are called from C and so require extra magic to be available at build time; see the comments in c-deps/libroach/db.cc.)

I also took the opportunity to clean up our header layout: public headers (i.e., those exposed to Go) now live in c-deps/libroachpb/include, while private headers live alongside their C++ source.

Besides the immediate win, there are some nice side-effects:

   * Undefined symbols properly cause compile-time errors instead of runtime panics. `-Wl,undefined` was too large a hammer, and could mask actual undefined symbol errors.

   * We can directly unit test our C and C++ code. Cgo previously made this nearly impossible: *_test.go files could not use cgo, so writing tests in Go was difficult, and the C and C++ files could not be compiled outside of cgo, so writing tests in C and C++ was difficult.

    * The parallel-cgo patch becomes good and truly unnecessary.

Fixes #14673.

benesch committed:
storage/engine: extract static library to fix windows ccl builds

Since the start of our Windows support, we've been unable to build CCL
binaries because ccl/storageccl/engineccl depended upon symbols from
storage/engine. These symbols were not provided when engineccl was
built, but rather when the final cockroach binary was linked. On
Unix-like platforms, we worked around the issue by passing
`-Wl,undefined` or its equivalent to ignore undefined symbols. Windows,
however, had no equivalent flag (the Portable Executable format that
Windows uses requires all symbols to be resolved at link time), and so
we were forced to omit the broken engineccl package by building the OSS
binary.

This commit extracts the C and C++ code from engine and engineccl into
their own static libraries, libroach and libroachccl, that are built
alongside our third-party C and C++ dependencies. This means engineccl
simply links against libroach *and* libroachccl so that all symbols are
available at build time. (Actually, there are two Go functions in
storage/engine that are called from C and so require extra magic to be
available at build time; see the comments in c-deps/libroach/db.cc.)

I also took the opportunity to clean up our header layout: public
headers (i.e., those exposed to Go) now live in c-deps/libroach/include,
while private headers live alongside their C++ source.

Besides the immediate win, there are some nice side-effects:

    * Undefined symbols properly cause compile-time errors instead of
      runtime panics. `-Wl,undefined` was too large a hammer, and could
      mask actual undefined symbol errors.

    * We can directly unit test our C and C++ code. Cgo previously made
      this nearly impossible: *_test.go files could not use cgo, so
      writing tests in Go was difficult, and the C and C++ files could
      not be compiled outside of cgo, so writing tests in C and C++ was
      difficult.

    * The parallel-cgo patch becomes good and truly unnecessary.

Fixes #14673.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17035)
<!-- Reviewable:end -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> :tada: 

---

Reviewed 64 of 64 files at r1.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[build/protobuf.mk, line 22 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7ZiP-YojUA94_zGnH:-Kp7ZiP-YojUA94_zGnI:b-f7w9d7) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/build/protobuf.mk#L22)):*
> ```Makefile
> include $(REPO_ROOT)/build/common.mk
> 
> NATIVE_ROOT := $(REPO_ROOT)/c-deps/roachlib/protos
> ```

shouldn't this be `libroach`? better yet, `LIBROACH_SRC_DIR`?

---

*[build/protobuf.mk, line 97 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7ZtHup49hzMNBFkHe:-Kp7ZtHup49hzMNBFkHf:b-cxkxjp) ([raw file](https://github.com/cockroachdb/cockroach/blob/e50b65991b5a8095103cc258965527030ab3dcff/build/protobuf.mk#L97)):*
> ```Makefile
> CPP_SOURCES_TARGET := $(LOCAL_BIN)/.cpp_protobuf_sources
> $(CPP_SOURCES_TARGET): $(PROTOC) $(CPP_PROTOS)
> 	(cd $(REPO_ROOT) && git ls-files --exclude-standard --cached --others -- '*.pb.h' '*.pb.cc' | xargs rm -f)
> ```

why did you remove this? we need to nuke all the generated code whenever we rebuild protos, otherwise we'll end up with generated code from deleted proto files.

---

*[c-deps/libroach-rebuild, line 4 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7_6oyjWQRgoFyPGU5:-Kp7_6oyjWQRgoFyPGU6:b-g3r94h) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/c-deps/libroach-rebuild#L4)):*
> ```
> ARTIFACT CACHING" in build/common.mk for rationale.
> 
> 2
> ```

why 2 and not 1?

---

*[c-deps/libroach/CMakeLists.txt, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7bfVoxiR2C64REpvT:-Kp7bfVoxiR2C64REpvU:bq5g4hx) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/c-deps/libroach/CMakeLists.txt#L1)):*
> ```CMake
> cmake_minimum_required(VERSION 3.1 FATAL_ERROR)
> ```

a comment explaining this minimum wouldn't hurt

---

*[c-deps/libroach/CMakeLists.txt, line 9 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7bZm-xQj2L20rfacM:-Kp7bZm-xQj2L20rfacN:b-4azgcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/c-deps/libroach/CMakeLists.txt#L9)):*
> ```CMake
>   encoding.cc
>   eventlistener.cc
>   protos/cockroach/pkg/roachpb/data.pb.cc
> ```

having to list these out is somewhat unfortunate.

---

*[c-deps/libroach/db.cc, line 46 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7aj_bkvRmvWiP5nvs:-Kp7aj_bkvRmvWiP5nvt:b9v575c) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/c-deps/libroach/db.cc#L46)):*
> ```cpp
> }
> 
> // These are Go functions exported by storage/engine. We provide these stubs,
> ```

you could also duplicate these functions into engineccl to solve this, right?

---

*[c-deps/libroach/ccl/db.cc, line 15 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7_Meev2fbL5J5GnrE:-Kp7_Meev2fbL5J5GnrF:b-21vxzb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/c-deps/libroach/ccl/db.cc#L15)):*
> ```cpp
> #include <rocksdb/utilities/write_batch_with_index.h>
> #include <libroach/ccl/db.h>
> #include "../db.h"
> ```

hm, why not `<libroach/db.h>`?

---

*[pkg/ccl/storageccl/engineccl/enable_cc.cc, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7__Rs5WAeUFH8Sz4D:-Kp7_WpomsANKSbJKKJO:b-trohk8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/pkg/ccl/storageccl/engineccl/enable_cc.cc#L1)):*
> ```cpp
> // This file exists solely to convince cgo to link this package against the C++
> ```

this comment should probably explain why this isn't done with ldflags (my guess is that it's because some platforms use libc++ and others use libstdc++)

---

*[pkg/ccl/storageccl/engineccl/rocksdb.go, line 23 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7_jEe3R_cBE0lherR:-Kp7_jEe3R_cBE0lherS:b-a1hvwv) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/pkg/ccl/storageccl/engineccl/rocksdb.go#L23)):*
> ```Go
> 
> // #cgo CPPFLAGS: -I../../../../c-deps/libroach/include
> // #cgo LDFLAGS: -lroachccl
> ```

does order matter? odd to see ccl first

---

*[pkg/storage/engine/enable_cc.cc, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7_n7wtBgHvvrY4JOm:-Kp7_n7wtBgHvvrY4JOn:b-trohk8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/pkg/storage/engine/enable_cc.cc#L1)):*
> ```cpp
> // This file exists solely to convince cgo to link this package against the C++
> ```

this comment should probably explain why this isn't done with ldflags (my guess is that it's because some platforms use libc++ and others use libstdc++)

---

*[pkg/storage/engine/rocksdb.go, line 65 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7_qRItwDCjpg9B39p:-Kp7_qRItwDCjpg9B39q:brp9wb0) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/pkg/storage/engine/rocksdb.go#L65)):*
> ```Go
> //
> // #include <stdlib.h>
> // #include "libroach/db.h"
> ```

why isn't this `<libroach/db.h>`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-:-Kp7cLyJpUw_qNZ__fT-:by6fmpn)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Did not review, but nice!

On Sat, Jul 15, 2017, 19:42 Tamir Duberstein <notifications@github.com>
wrote:

> [image: :lgtm:]
> <https://camo.githubusercontent.com/41b3ec3419116e3b3f84507ad965646ce4ce1f0c/68747470733a2f2f72657669657761626c652e696f2f6c67746d2e706e67>
> 🎉
> ------------------------------
>
> Reviewed 64 of 64 files at r1.
> Review status: all files reviewed at latest revision, 11 unresolved
> discussions.
> ------------------------------
>
> *build/protobuf.mk, line 22 at r1
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7ZiP-YojUA94_zGnH:-Kp7ZiP-YojUA94_zGnI:b-f7w9d7>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/build/protobuf.mk#L22>):*
>
> include $(REPO_ROOT)/build/common.mk
> NATIVE_ROOT := $(REPO_ROOT)/c-deps/roachlib/protos
>
> shouldn't this be libroach? better yet, LIBROACH_SRC_DIR?
> ------------------------------
>
> *build/protobuf.mk, line 97 at r1
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7ZtHup49hzMNBFkHe:-Kp7ZtHup49hzMNBFkHf:b-cxkxjp>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/e50b65991b5a8095103cc258965527030ab3dcff/build/protobuf.mk#L97>):*
>
> CPP_SOURCES_TARGET := $(LOCAL_BIN)/.cpp_protobuf_sources$(CPP_SOURCES_TARGET): $(PROTOC) $(CPP_PROTOS)
> 	(cd $(REPO_ROOT) && git ls-files --exclude-standard --cached --others -- '*.pb.h' '*.pb.cc' | xargs rm -f)
>
> why did you remove this? we need to nuke all the generated code whenever
> we rebuild protos, otherwise we'll end up with generated code from deleted
> proto files.
> ------------------------------
>
> *c-deps/libroach-rebuild, line 4 at r1
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7_6oyjWQRgoFyPGU5:-Kp7_6oyjWQRgoFyPGU6:b-g3r94h>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/c-deps/libroach-rebuild#L4>):*
>
> ARTIFACT CACHING" in build/common.mk for rationale.
>
> 2
>
> why 2 and not 1?
> ------------------------------
>
> *c-deps/libroach/CMakeLists.txt, line 1 at r1
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7bfVoxiR2C64REpvT:-Kp7bfVoxiR2C64REpvU:bq5g4hx>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/c-deps/libroach/CMakeLists.txt#L1>):*
>
> cmake_minimum_required(VERSION 3.1 FATAL_ERROR)
>
> a comment explaining this minimum wouldn't hurt
> ------------------------------
>
> *c-deps/libroach/CMakeLists.txt, line 9 at r1
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7bZm-xQj2L20rfacM:-Kp7bZm-xQj2L20rfacN:b-4azgcz>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/c-deps/libroach/CMakeLists.txt#L9>):*
>
>   encoding.cc
>   eventlistener.cc
>   protos/cockroach/pkg/roachpb/data.pb.cc
>
> having to list these out is somewhat unfortunate.
> ------------------------------
>
> *c-deps/libroach/db.cc, line 46 at r1
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7aj_bkvRmvWiP5nvs:-Kp7aj_bkvRmvWiP5nvt:b9v575c>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/c-deps/libroach/db.cc#L46>):*
>
> }
> // These are Go functions exported by storage/engine. We provide these stubs,
>
> you could also duplicate these functions into engineccl to solve this,
> right?
> ------------------------------
>
> *c-deps/libroach/ccl/db.cc, line 15 at r1
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7_Meev2fbL5J5GnrE:-Kp7_Meev2fbL5J5GnrF:b-21vxzb>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/c-deps/libroach/ccl/db.cc#L15>):*
>
> #include <rocksdb/utilities/write_batch_with_index.h>
> #include <libroach/ccl/db.h>
> #include "../db.h"
>
> hm, why not <libroach/db.h>?
> ------------------------------
>
> *pkg/ccl/storageccl/engineccl/enable_cc.cc, line 1 at r1
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7__Rs5WAeUFH8Sz4D:-Kp7_WpomsANKSbJKKJO:b-trohk8>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/pkg/ccl/storageccl/engineccl/enable_cc.cc#L1>):*
>
> // This file exists solely to convince cgo to link this package against the C++
>
> this comment should probably explain why this isn't done with ldflags (my
> guess is that it's because some platforms use libc++ and others use
> libstdc++)
> ------------------------------
>
> *pkg/ccl/storageccl/engineccl/rocksdb.go, line 23 at r1
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7_jEe3R_cBE0lherR:-Kp7_jEe3R_cBE0lherS:b-a1hvwv>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/pkg/ccl/storageccl/engineccl/rocksdb.go#L23>):*
>
> // #cgo CPPFLAGS: -I../../../../c-deps/libroach/include// #cgo LDFLAGS: -lroachccl
>
> does order matter? odd to see ccl first
> ------------------------------
>
> *pkg/storage/engine/enable_cc.cc, line 1 at r1
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7_n7wtBgHvvrY4JOm:-Kp7_n7wtBgHvvrY4JOn:b-trohk8>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/pkg/storage/engine/enable_cc.cc#L1>):*
>
> // This file exists solely to convince cgo to link this package against the C++
>
> this comment should probably explain why this isn't done with ldflags (my
> guess is that it's because some platforms use libc++ and others use
> libstdc++)
> ------------------------------
>
> *pkg/storage/engine/rocksdb.go, line 65 at r1
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7_qRItwDCjpg9B39p:-Kp7_qRItwDCjpg9B39q:brp9wb0>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/pkg/storage/engine/rocksdb.go#L65>):*
>
> //// #include <stdlib.h>// #include "libroach/db.h"
>
> why isn't this <libroach/db.h>?
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-:-Kp7cLyJpUw_qNZ__fT-:by6fmpn>*
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/17035#issuecomment-315570754>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135K4K4CcBEgU8Y-5129h9jw_WLAUMks5sOU5jgaJpZM4OZJsl>
> .
>
-- 

-- Tobias

benesch commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[build/protobuf.mk, line 22 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7ZiP-YojUA94_zGnH:-Kp8ovuioJ8McYNt7mlN:bp14ag8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/build/protobuf.mk#L22)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

shouldn't this be `libroach`? better yet, `LIBROACH_SRC_DIR`?
</blockquote></details>

Eek, thanks. Thought I fixed that.

---

*[build/protobuf.mk, line 97 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7ZtHup49hzMNBFkHe:-Kp8oxwrXMtQdtmgTKt8:bcksf83) ([raw file](https://github.com/cockroachdb/cockroach/blob/e50b65991b5a8095103cc258965527030ab3dcff/build/protobuf.mk#L97)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why did you remove this? we need to nuke all the generated code whenever we rebuild protos, otherwise we'll end up with generated code from deleted proto files.
</blockquote></details>

Ah, you're right. Thought it was part of the symlink code. Restored.

---

*[c-deps/libroach-rebuild, line 4 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7_6oyjWQRgoFyPGU5:-Kp8p057UyOueiqhFxaU:b7r0550) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/c-deps/libroach-rebuild#L4)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why 2 and not 1?
</blockquote></details>

Because I copied it from protobuf-rebuild. Fixed. Thanks!

---

*[c-deps/libroach/CMakeLists.txt, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7bfVoxiR2C64REpvT:-Kp8qDu-SIBhl8FyCNEP:bka1e4g) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/c-deps/libroach/CMakeLists.txt#L1)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

a comment explaining this minimum wouldn't hurt
</blockquote></details>

This works as far back as the documentation goes (3.0.2)—based on my reading of the docs—but I don't have an easy way of testing that. Thoughts? I just picked what our documentation demands (for Snappy).

---

*[c-deps/libroach/CMakeLists.txt, line 9 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7bZm-xQj2L20rfacM:-Kp8pBvpxQp5Rw-423qQ:bul01zo) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/c-deps/libroach/CMakeLists.txt#L9)):*
There is an alternative that uses `file(GLOB...)`, but from the CMake docs:

> We do not recommend using GLOB to collect a list of source files from your source tree. If no CMakeLists.txt file changes when a source is added or removed then the generated build system cannot know when to ask CMake to regenerate.

Basically, CMake isn't smart enough to realize that the generated Makefile is out of date when you simply add another file to a directory. It's better to force developers to explicitly list neww files in `CMakeLists.txt` so that the build system is auto-regenerated.

---

*[c-deps/libroach/db.cc, line 46 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7aj_bkvRmvWiP5nvs:-Kp8pe8eKm96BkTTReWy:bujm0ir) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/c-deps/libroach/db.cc#L46)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

you could also duplicate these functions into engineccl to solve this, right?
</blockquote></details>

Yup, but I'd rather avoid the duplicated code. PE supports weak symbols, fortunately.

---

*[c-deps/libroach/ccl/db.cc, line 15 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7_Meev2fbL5J5GnrE:-Kp8oVR8dSzP6Xwol-9f:bhy0wlb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/c-deps/libroach/ccl/db.cc#L15)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

hm, why not `<libroach/db.h>`?
</blockquote></details>

`<libroach/db.h>` is the public header; this needs to explicitly include the private header, which declares functions that should not be available to Go.

---

*[pkg/ccl/storageccl/engineccl/enable_cc.cc, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7__Rs5WAeUFH8Sz4D:-Kp8oRRyeSmPcSIM7OH_:b9izszt) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/pkg/ccl/storageccl/engineccl/enable_cc.cc#L1)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this comment should probably explain why this isn't done with ldflags (my guess is that it's because some platforms use libc++ and others use libstdc++)
</blockquote></details>

Yup, done.

---

*[pkg/ccl/storageccl/engineccl/rocksdb.go, line 23 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7_jEe3R_cBE0lherR:-Kp8oI5JI6lV-Uu-BJXw:bkpz47z) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/pkg/ccl/storageccl/engineccl/rocksdb.go#L23)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

does order matter? odd to see ccl first
</blockquote></details>

Yup, order matters. The library that depends on a symbol must be listed before the library that provides it.

---

*[pkg/storage/engine/enable_cc.cc, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7_n7wtBgHvvrY4JOm:-Kp8oFcp6u9tkVvAHg5z:b9izszt) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/pkg/storage/engine/enable_cc.cc#L1)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this comment should probably explain why this isn't done with ldflags (my guess is that it's because some platforms use libc++ and others use libstdc++)
</blockquote></details>

Yup, done.

---

*[pkg/storage/engine/rocksdb.go, line 65 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7_qRItwDCjpg9B39p:-Kp8n73HaWOpuvdPMvL5:b-r6jnys) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/pkg/storage/engine/rocksdb.go#L65)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

why isn't this `<libroach/db.h>`?
</blockquote></details>

Oversight, thanks! Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 58 of 64 files at r1, 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[build/common.mk, line 351 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp9tYuMHpoNl7-4o7AZ:-Kp9tYuMHpoNl7-4o7A_:b-39r6hv) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f024c645c00db1087835d2d163923ef49ed0d04/build/common.mk#L351)):*
> ```Makefile
> #
> # NB: the recipes below nuke *all* build artifacts when a dependency's source
> # tarball changes. Normally, we could rely on the dependency's build system to
> ```

I *think* this is last paragraph is stale.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035)*
<!-- Sent from Reviewable.io -->

benesch commented:
Review status: 37 of 41 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[build/common.mk, line 351 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp9tYuMHpoNl7-4o7AZ:-KpDVpWmvAOMwKFOskCi:b-whf8l1) ([raw file](https://github.com/cockroachdb/cockroach/blob/8f024c645c00db1087835d2d163923ef49ed0d04/build/common.mk#L351)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I *think* this is last paragraph is stale.
</blockquote></details>

Ooh, good call. Will address in a separate PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 55 of 64 files at r1, 4 of 6 files at r2, 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[c-deps/libroach/db.cc, line 46 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7aj_bkvRmvWiP5nvs:-KpFzTS6Lt76VkBY-rtY:b-h3h4vp) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/c-deps/libroach/db.cc#L46)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Yup, but I'd rather avoid the duplicated code. PE supports weak symbols, fortunately.
</blockquote></details>

You wouldn't have to duplicate the code, just the symbols, right? 

---

*[c-deps/libroach/db.h, line 21 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-KpFziemywwUpdN7Rorq:-KpFziemywwUpdN7Rorr:b7pt7sz) ([raw file](https://github.com/cockroachdb/cockroach/blob/16a552f8caa2f3ceed4012c0e2443606836c1604/c-deps/libroach/db.h#L21)):*
> ```objc
> #include <rocksdb/write_batch.h>
> #include <rocksdb/write_batch_base.h>
> #include <libroach/db.h>
> ```

This file is also named `libroach/db.h`. Better to keep the `db_internal.h` name to avoid confusion (or some other renaming).

---

*[c-deps/libroach/include/libroach/ccl/db.h, line 30 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-KpG0rpMNL9NAK3CF0oq:-KpG0rpMNL9NAK3CF0or:b-nqm21s) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf5924554700ee5c779bd13ab6cd53188871f677/c-deps/libroach/include/libroach/ccl/db.h#L30)):*
> ```objc
> 
> // local variables:
> // mode: c++
> ```

Either leave this in or remove it from the non-ccl header too. (I think remove it - this file is C, not C++, with the exception of the extern C lines. `db_internal.h` could use it, although it would go against our preference not to clutter the repo with editor-specific cruft)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-:-KpG1znUywaj4nnfD5VM:bnfp4nl)*
<!-- Sent from Reviewable.io -->

danhhz commented:
I'd be entirely sciencedogging the review, but I'm in favor of making this happen. Good work!

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[c-deps/libroach/ccl/db.cc, line 1 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-KpGSi10WVPL1JNlEQM9:-KpGSi10WVPL1JNlEQMA:b7bxqtz) ([raw file](https://github.com/cockroachdb/cockroach/blob/16a552f8caa2f3ceed4012c0e2443606836c1604/c-deps/libroach/ccl/db.cc#L1)):*
> ```cpp
> // Copyright 2017 The Cockroach Authors.
> ```

we've been assuming that _all_ ccl code lives under pkg/ccl. can you update pkg/ccl/README.md to mention this new dir?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-:-KpGSYMRSs4E3EegAda5:b-j15bma)*
<!-- Sent from Reviewable.io -->

benesch commented:
@petermattis did you want to take a look at this, or was your LGTM on #17040 transitive?

benesch commented:
Review status: 35 of 43 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[c-deps/libroach/db.cc, line 46 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7aj_bkvRmvWiP5nvs:-KpGRsgQb4qB-Rz8VUGP:bt309vz) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/c-deps/libroach/db.cc#L46)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

You wouldn't have to duplicate the code, just the symbols, right? 
</blockquote></details>

In theory—and I haven't actually tried—but I think the way cgo generates types makes it essentially impossible to duplicate just the symbols:

https://github.com/cockroachdb/cockroach/blob/master/pkg/ccl/storageccl/engineccl/rocksdb.go#L94

---

*[c-deps/libroach/db.h, line 21 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-KpFziemywwUpdN7Rorq:-KpH-jrHu_95dTSVopJq:bfb6cc1) ([raw file](https://github.com/cockroachdb/cockroach/blob/16a552f8caa2f3ceed4012c0e2443606836c1604/c-deps/libroach/db.h#L21)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This file is also named `libroach/db.h`. Better to keep the `db_internal.h` name to avoid confusion (or some other renaming).
</blockquote></details>

Good point. Given that we only have one non-CCL and one CCL header file, I've renamed them `libroach.h` and `libroachccl.h`. That way this one continues to be "db.h", matching its C++ source file.

---

*[c-deps/libroach/ccl/db.cc, line 1 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-KpGSi10WVPL1JNlEQM9:-KpH-wqUY7iFaBxhF6vg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/16a552f8caa2f3ceed4012c0e2443606836c1604/c-deps/libroach/ccl/db.cc#L1)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

we've been assuming that _all_ ccl code lives under pkg/ccl. can you update pkg/ccl/README.md to mention this new dir?
</blockquote></details>

Done.

---

*[c-deps/libroach/include/libroachccl.h, line 30 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-KpG0rpMNL9NAK3CF0oq:-KpH018rrM9jMZqUw7Vt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf5924554700ee5c779bd13ab6cd53188871f677/c-deps/libroach/include/libroachccl.h#L30)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Either leave this in or remove it from the non-ccl header too. (I think remove it - this file is C, not C++, with the exception of the extern C lines. `db_internal.h` could use it, although it would go against our preference not to clutter the repo with editor-specific cruft)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035)*
<!-- Sent from Reviewable.io -->

petermattis commented:
I'm fine with doing this, though I'll have to retrain my fingers to locate `db.c`. I'll leave the details of the review to others.

---

Review status: 35 of 43 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-:-KpH3AvEp7SZVAgURZ0L:b64zvp2)*
<!-- Sent from Reviewable.io -->

benesch commented:
Review status: 35 of 43 files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[c-deps/libroach/db.cc, line 46 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035#-Kp7aj_bkvRmvWiP5nvs:-KpH3gM4vjrc7d1Tn95h:b-8fo2vb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f5cf4b9779ce0fd3a55664c9559dee546535ba1b/c-deps/libroach/db.cc#L46)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

In theory—and I haven't actually tried—but I think the way cgo generates types makes it essentially impossible to duplicate just the symbols:

https://github.com/cockroachdb/cockroach/blob/master/pkg/ccl/storageccl/engineccl/rocksdb.go#L94
</blockquote></details>

Yup, verified. https://github.com/golang/go/issues/13467

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035)*
<!-- Sent from Reviewable.io -->

benesch commented:
Ok, in that case, I'll merge since @bdarnell and @tamird are happy. TFTRs!

tamird commented:
Reviewed 5 of 66 files at r1, 3 of 5 files at r3, 10 of 10 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17035)*
<!-- Sent from Reviewable.io -->
